src/main/java/com/google/sps/data/DeliverySystem.java

```diff
@@ -26,16 +26,74 @@
+    Journey journey = findOrdersForDeliveryRequest(deliverySlot);
+    return journey.findOptimalOrderForWaypoints();
+  }
+
+  /**
+   * Assigns a set of orders for the delivery request such that they fit the [startTime, endTime]
+   * timeslot.
+   */
+  public Journey findOrdersForDeliveryRequest(DeliverySlot deliverySlot)  throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
+    Point startPoint = deliverySlot.getStartPoint();
+    Journey journey = new Journey(startPoint, pathFinder);
+    OrderHandler orderHandler = new OrderHandler(pathFinder);
+     // Get unassigned orders from datastore which are in the area of the starting point.
+    List<Entity> orders = orderHandler.getAvailableOrders(startPoint.getArea());
+    for (Entity order : orders) {
+      if (journey.getNumberOfWaypoints() >= MAX_WAYPOINTS) {
+        break;
+      }
+
+      LibraryPoint library = new LibraryPoint((double) order.getProperty(OrderHandler.OrderProperty.LIBRARY_LAT.label), 
+        (double) order.getProperty(OrderHandler.OrderProperty.LIBRARY_LNG.label),
+        (int) order.getProperty(OrderHandler.OrderProperty.LIBRARY_ID.label));
+      Point recipient = new Point((double) order.getProperty(OrderHandler.OrderProperty.RECIPIENT_LAT.label), 
+        (double) order.getProperty(OrderHandler.OrderProperty.RECIPIENT_LNG.label));
+      
+      // Check if the library is already a waypoint in the journey, and add it in case it's not.
+      boolean libraryIsWaypoint = !journey.addPointToWaypoints(library);
+      // Check if the recipient is already a waypoint in the journey, and add it in case it's not.
+      boolean recipientIsWaypoint = !journey.addPointToWaypoints(recipient);
+
+      if (journey.getNumberOfWaypoints() > MAX_WAYPOINTS || !journey.findJourneyForTimeslot(deliverySlot)) {
+        // If the new journey doesn't satisfy user's prefrences, remove the order.
+        if (!libraryIsWaypoint) {
+          // Remove the library from waypoints if it was added only for this order.
+          journey.removeWaypoint(library);
+        }
+        if (!recipientIsWaypoint) {
+          // Remove the recipient from the waypoints if he has no other order assigned to this 
+          // delivery slot.
+          journey.removeWaypoint(recipient);
+        }
+      } else {
+        // The library must be visited before the recipient.
+        journey.addRestriction(journey.getWaypointIndex(library), journey.getWaypointIndex(recipient));
+        // Add the books from current order to the library their are assigned to.
+        library.addBooks((List<String>)order.getProperty("books"));
+        // The order is assigned to the delivery slot and its status changes in datastore.
+        order.setProperty("status", "ASSIGNED");
```

Comment:
You'll probably need to store a lot more data here, for each waypoint you need to store what order ids to pick up or deliver, so you can actually display some that data to the courier.

Reply:
Yes, for each waypoint, I store a private list with the IDs of the books that must be borrowed/delivered.

Reply:
Do you? I see this in Journey:
  private List<Point> waypoints;
  private List<String> orderKeys;


But I don't see how you can infer from that which orders belong to which waypoint?

Reply:
I added them here: https://github.com/googleinterns/step235-2020/pull/33.